starboard/nplb/maximum_player_configuration_explorer_test.cc

```diff
@@ -0,0 +1,168 @@
+       MaximumPlayerConfigurationExplorerTest) {
+  MaximumPlayerConfigurationExplorer explorer(player_configs_,
+                                              kMaxPlayerInstancesPerConfig,
+                                              &fake_graphics_context_provider_);
+  auto max_config_set = explorer.CalculateMaxTestConfigSet();
+  EXPECT_GT(max_config_set.size(), 0);
+  for (const auto& max_config : max_config_set) {
+    for (const auto& v : max_config) {
+      EXPECT_GE(kMaxPlayerInstancesPerConfig, v);
+    }
+  }
+}
+
+TEST_F(MaximumPlayerConfigurationExplorerTest,
+       MaximumPlayerConfigurationExplorerTestWithLimitationInTotalInstances) {
+  auto total_instance_limit_function =
+      std::bind(CheckTotalResourcesExceedLimitation, std::placeholders::_1,
+                kMaxTotalPlayerInstances);
+
+  MaximumPlayerConfigurationExplorer explorer(
+      player_configs_, kMaxPlayerInstancesPerConfig,
+      &fake_graphics_context_provider_, total_instance_limit_function);
+  auto max_config_set = explorer.CalculateMaxTestConfigSet();
+  EXPECT_GT(max_config_set.size(), 0);
+  for (const auto& max_config : max_config_set) {
+    int total_resources = 0;
+    for (const auto& v : max_config) {
+      EXPECT_GE(kMaxPlayerInstancesPerConfig, v);
+      total_resources += v;
+    }
+    EXPECT_GE(kMaxTotalPlayerInstances, total_resources);
+  }
+
+  auto multiple_player_test_config = GetMultiplePlayerTestConfig(
+      kSbPlayerOutputModeDecodeToTexture, kMaxPlayerInstancesPerConfig,
+      kMaxTotalPlayerInstances, &fake_graphics_context_provider_);
+  EXPECT_EQ(multiple_player_test_config.size(), max_config_set.size());
+}
+
+// Verify whether all the immediate children of the maximum elements set are
+// also supportable. For example, when resource types = 3 and max instances = 5,
+// if the maximum elements set = [5, 5, 5], we will check {[4, 5, 5], [5, 4, 5],
+// [5, 5, 4]} is also valid here.
+TEST_F(MaximumPlayerConfigurationExplorerTest,
+       MaximumPlayerConfigurationExplorerTestImmediateChildren) {
+  // Calculate the maximum elements set firstly.
+  std::unique_ptr<MaximumPlayerConfigurationExplorer> explorer(
+      new MaximumPlayerConfigurationExplorer(player_configs_,
+                                             kMaxPlayerInstancesPerConfig,
+                                             &fake_graphics_context_provider_));
+  auto max_config_set = explorer->CalculateMaxTestConfigSet();
+
+  // Collect the immediate children of the maximum elements set.
+  std::set<std::vector<int>> immediate_children_of_maximum_elements;
+  for (auto& max_config : max_config_set) {
+    for (int i = 0; i < max_config.size(); ++i) {
+      if (max_config[i] > 0) {
+        auto new_max_config = max_config;
+        --new_max_config[i];
+        immediate_children_of_maximum_elements.insert(new_max_config);
+      }
+    }
+  }
+  if (immediate_children_of_maximum_elements.empty()) {
+    return;
+  }
```

Comment:
Is this check necessary? i.e. can MaximumPlayerConfigurationExplorer work if validate_function always returns false?

Reply:
Yes, it is possible that there is no configuration supported at all. However, this is almost impossible for Cobalt. For example, if there is a very old device that can only play Motion JPEG, then Cobalt would not be able to play any video on that device.
In this case immediate_children_of_maximum_elements.empty() = true.
Notice that it implies max_config_set = {(0, 0, .. 0)} but immediate_children_of_maximum_elements = {}

If we run MaximumPlayerConfigurationExplorer with {} validate_function, MaximumPlayerConfigurationExplorer will retunr (0, 0, ...0) when it supports nothing.

Since {(0, 0, .. 0)}  != {}, we will fail at
```
EXPECT_TRUE(max_config_set == immediate_children_of_maximum_elements);
```

Reply:
That makes sense.  Let's add a brief comments here.  It shouldn't be the detailed explanation you had above, something like "the ... is empty, but CalculateMaxTestConfigSet() will return a config set like {0, 0, 0, ..., 0} and fails the EXPECT_TRUE below" is good enough.